ls1intum/Artemis
# [Enhancements] - Exam statistics visualizations

### Checklist

### Motivation and Context


We want to extend the information shown on the exam scores page. And all the tables look a bit boring, we want to spice it up by adding some color and some graphs 🌶️ 

### Description

- Added additional columns which are only shown when there is a second correction round and this second correction round already started, so someone assessed a submission a second time
- This additional column gives information regarding the first correction so instructors can identify how the second correction influenced the outcome of the exam
- Added a horizontal bar chart for every exercise group, which displays the average score for the exercise at first (blue bar) and each exercise (dark bar). Those charts visualize the information given in the table below

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Conduct an exam with second correction
4. While assessing the first correction, make sure that when looking at the scores page of the exam, no additional table entries are shown (so should look like 1. of screenshots)
5. During and after second assessment, the table in the scores page on top of the page also shows information regarding the first correction round
6. Below the table there are horizontal bar charts which visualize the correct average score (and points when hovering over a bar) - compare thsoe values to the average scores shown in the table below

### Test Coverage



exam-scores-average-scores-graph.component.ts: 98%
improved exam-scores.component.ts to: 94%

### Screenshots



Exam without second correction and without grade key, remains unchanged:
![Screenshot 2021-06-07 at 16 37 32](https://user-images.githubusercontent.com/44401560/121036189-dc592a00-c7ae-11eb-9ca0-d773c6e4b2d6.png)
 
Exam with second correction but without grade key:
![Screenshot 2021-06-07 at 16 37 17](https://user-images.githubusercontent.com/44401560/121036297-f2ff8100-c7ae-11eb-8237-1f4900ee26bf.png)

Table of an exam with second correction and grade key
![Screenshot 2021-06-07 at 16 37 01](https://user-images.githubusercontent.com/44401560/121036119-cea3a480-c7ae-11eb-8623-5c03b83850a9.png)

The horizontal bar charts visualizing the table below:


<img width="841" alt="Screenshot 2021-06-09 at 18 57 43" src="https://user-images.githubusercontent.com/44401560/121398106-490e2900-c955-11eb-9196-75a5624a0882.png">
